### PR TITLE
Improved building of requests with body

### PR DESCRIPTION
**Pull request by tomtit**

### CODE REVIEW

File: readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java [hunk 1]
```diff
@@ -91,31 +84,87 @@ private Swagger parseSwagger(String swaggerUrl) throws IllegalArgumentException
...
+    private void ensureBodyParameter(Operation op) {
+        if (op.getParameters().stream().noneMatch(param -> param instanceof BodyParameter)) {
+            throw new IllegalArgumentException(
+                    String.format("Body parameter is not defined for the [%s] operation",
+                            op.getOperationId()));
+        }
+    }
+
+    private void ensureHttpMethodWithBody(TestSteps.HttpMethod method) {
+        if (!Arrays.asList(TestSteps.HttpMethod.POST, TestSteps.HttpMethod.PUT,
```
Comment (olensmar) [inline readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java]: can we make this list static so it isn't created every time?

File: readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java [hunk 2]
```diff
@@ -91,31 +84,87 @@ private Swagger parseSwagger(String swaggerUrl) throws IllegalArgumentException
...
+
+    /**
+     * Creates a RestRequestStepWithBodyBuilder for an arbitrary path and method - the
+     * targetEndpoint and basePath will be prefixed to the specified path.
+     *
+     * @param path   the path to append to the targetEndpoint and basePath
+     * @param method the HTTP method to use
+     */
+    public RestRequestStepWithBodyBuilder requestWithBody(String path, TestSteps.HttpMethod method){
+        String basePath = swagger.getBasePath();
```
Comment (olensmar) [inline readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java]: suggest we extract this logic to get the basePath into a separate method and call it from here and above

File: readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java [hunk 3]
```diff
@@ -136,6 +172,14 @@ public String getTargetEndpoint() {
      * @param targetEndpoint the target endpoint of the API under test
      */
     public void setTargetEndpoint(String targetEndpoint) {
-        this.targetEndpoint = targetEndpoint;
+        final String endpoint = Optional.ofNullable(targetEndpoint).orElseGet(() -> {
```
Comment (olensmar) [inline readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java]: i feel this is overdoing the streams/lambda thing :-)

Comment (spindelmanne) [inline readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java]: +1 :-)

Comment (olensmar) [inline readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java]: please revert to some good old if/else statements :-)

File: readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java [hunk 4]
```diff
@@ -91,31 +87,71 @@ private Swagger parseSwagger(String swaggerUrl) throws IllegalArgumentException
...
+     * @param method the HTTP method to use
+     */
+    public RestRequestStepWithBodyBuilder requestWithBody(String path, TestSteps.HttpMethod method){
+        ensureHttpMethodWithBody(method);
+        return new RestRequestStepWithBodyBuilder(targetBasePath + path, method);
+    }
 
-        if( isHttpMethodWithBody( method )){
-            return new RestRequestStepWithBodyBuilder(targetEndpoint + basePath + path, method);
+    private void ensureBodyParameter(Operation op) {
```
Comment (spindelmanne) [inline readyapi4j-swagger/src/main/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilder.java]: Rename the parameter to operation.

File: readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java [hunk 5]
```diff
@@ -29,8 +36,37 @@ public void testExistingOperation() throws Exception {
     public void testMissingOperation() throws Exception {
         try {
             petstore.operation("tjoho");
-            assertFalse("Expected IllegalArgumentException for invalid operationId", true);
+            fail("Expected IllegalArgumentException for invalid operationId");
```
Comment (spindelmanne) [inline readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java]: I missed this when reviewing Ole's code so I complain now instead ...

The cleanest and most readable way to verify this is to use the expected attribute in the @Test annotation:

@Test(expected = IllegalArgumentException.class)

Comment (olensmar) [inline readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java]: thanks - wasn't aware of that!

Comment (tomtit) [inline readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java]: What is the right way to verify error message?

**Comment by tomtit** [comment]
Welcome Java 1.0 :)

File: readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java (hunk 5 above)
Comment (spindelmanne) [inline readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java]: You can use the @Rule annotation to do that in a more elegant way. See here: http://blog.codeleak.pl/2013/07/3-ways-of-handling-exceptions-in-junit.html

Comment (tomtit) [inline readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java]: Done

**Comment by olensmar** [comment]
looks good to me now - @spindelmanne ?

File: readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java [hunk 6]
```diff
@@ -27,11 +37,31 @@ public void testExistingOperation() throws Exception {
...
+        thrown.expectMessage("operationId [tjoho] not found in Swagger definition");
+        petstore.operation("tjoho");
+    }
+
+    @Test
+    public void testOperationWithBody() {
+        final RestTestRequestStep step = petstore.operationWithBody("addPet")
+                .withRequestBody("body").build();
+        assertEquals("body", step.getRequestBody());
+
```
Comment (spindelmanne) [inline readyapi4j-swagger/src/test/java/com/smartbear/readyapi4j/swagger/SwaggerTestStepBuilderTest.java]: Unnecessary whitespace, but let's just remove that later.

**Pull request merged**